hermes-common/src/main/java/pl/allegro/tech/hermes/domain/topic/schema/SchemaRepository.java

```diff
@@ -0,0 +1,114 @@
+    private final SchemaCompiler<T> schemaCompiler;
+
+    @Inject
+    public SchemaRepository(SchemaSourceProvider schemaRepository, ExecutorService reloadSchemaSourceExecutor, SchemaCompiler<T> schemaCompiler) {
+        this(schemaRepository, reloadSchemaSourceExecutor, Ticker.systemTicker(), schemaCompiler);
+    }
+
+    SchemaRepository(SchemaSourceProvider schemaSourceProvider, ExecutorService reloadSchemaSourceExecutor, Ticker ticker, SchemaCompiler<T> schemaCompiler) {
+        this.schemaCompiler = schemaCompiler;
+        this.schemaCache = CacheBuilder
+                .newBuilder()
+                .ticker(ticker)
+                .refreshAfterWrite(10, TimeUnit.MINUTES)
+                .expireAfterWrite(24, TimeUnit.HOURS)
+                .build(new CacheLoader<Topic, SchemaWithSource>() {
+                    @Override
+                    public SchemaWithSource load(Topic topic) throws Exception {
+                        SchemaSource newRawSource = findSchemaSource(topic);
+                        logger.info("Loading schema for topic {}", topic.getQualifiedName());
+                        return createSchemaWithSource(newRawSource);
+                    }
+
+                    @Override
+                    public ListenableFuture<SchemaWithSource> reload(Topic topic, SchemaWithSource oldSchemaWithSource) throws Exception {
+                        SchemaSource newRawSource;
+                        try {
+                            newRawSource = findSchemaSource(topic);
+                        } catch (Exception e) {
+                            logger.warn("Could not reload schema for topic {}", topic.getQualifiedName(), e);
+                            return Futures.immediateFuture(oldSchemaWithSource);
+                        }
+
+                        if (oldSchemaWithSource.getSource().equals(newRawSource)) {
+                            return Futures.immediateFuture(oldSchemaWithSource);
+                        }
+
+                        ListenableFutureTask<SchemaWithSource> task = ListenableFutureTask.create(() -> {
+                            logger.info("Reloading schema for topic {}", topic.getQualifiedName());
+                            try {
+                                return createSchemaWithSource(newRawSource);
+                            } catch (Exception e) {
+                                logger.warn("Could not compile schema for topic {}", topic.getQualifiedName(), e);
+                                throw e;
+                            }
+                        });
+                        reloadSchemaSourceExecutor.execute(task);
+                        return task;
+                    }
+
+                    private SchemaSource findSchemaSource(Topic topic) {
+                        return schemaSourceProvider.get(topic).orElseThrow(() -> new SchemaSourceNotFoundException("No schema source for topic " + topic.getQualifiedName()));
```

Comment:
you could extract "No schema source for topic " to SchemaSourceNotFoundException internals and pass only qualifiedTopicName